source/docs/software/commandbased/organizing-command-based.rst

```diff
@@ -0,0 +1,259 @@
+
+A command like the intake-running command is conceptually related to exactly one subsystem: the ``Intake``. As such, it makes sense to put a ``commandRun`` method as an instance method of the ``Intake`` class:
+
+.. tabs::
+
+  .. code-tab:: java
+
+    public class Intake {
+        // [code for motor controllers, configuration, etc.]
+        // ...
+
+        public Command commandRun() {
+            return new StartEndCommand(() -> this.set(1), () -> this.set(0), this);
+        }
+    }
+
+  .. code-tab:: c++
+
+    // TODO
+
+Notice how since we are in the ``Intake`` class, we no longer refer to ``intake``; instead, we use the ``this`` keyword to refer to the current instance.
+
+.. warning:: To preserve encapsulation, avoid referring to ``private`` variables and methods of the ``Intake`` class from within the ``commandRun`` method, even though they are technically accessible. (For instance, the ``commandRun`` method should use the subsystem's public ``set`` method instead of interfacing directly with the motor controllers.)
+
+Using this new factory method in command groups and button bindings is highly expressive:
+
+
+.. tabs::
+
+  .. code-tab:: java
+
+    intakeButton.whileHeld(intake.commandRun());
+
+    Command intakeAndShoot = intake.commandRun().alongWith(new RunShooter());
+
+    Command autonomousCommand = new SequentialCommandGroup(
+        intake.commandRun().withTimeout(5),
+        new WaitCommand(3),
+        intake.commandRun().withTimeout(5)
+    );
+
+  .. code-tab:: c++
+
+    // TODO
+
+Adding a parameter to the ``commandRun`` method to provide the exact percentage to run the intake is easy and allows for even more flexibility.
+
+.. tabs::
+
+  .. code-tab:: java
+
+    public Command commandRun(double percent) {
+        return new StartEndCommand(() -> this.set(percent), () -> this.set(0), this);
+    }
+
+  .. code-tab:: c++
+
+    // TODO
+
+For instance, this code creates a command group that runs the intake forwards for two seconds, waits for two seconds, and then runs the intake backwards for five seconds.
+
+.. tabs::
+
+  .. code-tab:: java
+
+    Command intakeRunSequence = intake.commandRun(1).withTimeout(2)
+        .andThen(new WaitCommand(2))
+        .andThen(intake.commandRun(-1).withTimeout(5));
+
+  .. code-tab:: c++
+
+    // TODO
+
+Command Groups
+^^^^^^^^^^^^^^
+
+Command groups have slightly different organizational concerns, but many of the same principles apply. Certain command groups are used in just one place, and so it's more sensible to define a command group entirely using inline decorators shortly before it is used (such as the ``intakeRunSequence`` command immediately above). However, command groups that are reused often, or large command groups such as autonomous routines, are still better when split into separate files. We'll consider command group using the same example from the previous section:
```

Comment:
I'm not sure about this split. I think that the line between command and group is very blurry, and often times the "simple" factories in the subsystem are groups as well--that isn't necessarily clear from the way this is currently written.

I think the decision of whether the command should be in a separate file should be based on number of requirements rather than "how much it's used".

Reply:
Maybe this prose is unclear—by "separate file" I mean defined right when it's used, for example in `RobotContainer` as button bindings are being set up. I definitely agree that the line between commands and command groups is a bit awkward though.

I'm trying to preserve a sort of narrative, starting with very simple intake commands and moving on to command groups in separate static classes. If this were a spec sheet of information it's probably better to use the four different categories, but I'm not sure if that's best here.

Reply:
Most of the articles (especially the command-based ones) are spreads of information, and the narrative/progression is only at the article level (ie articles about simpler topics are before articles about advanced topics).

Each of the structures has a use case that it is good for (and the example code should show that), so sorting the sections by the complexity of the example/use case is a good idea.
I'd say something like this:
1. instance factory methods -- simple intake commands
2. subclassing CommandBase  -- single commands that have somewhat custom logic (using the same example as 1 to show how it can get verbose is a good idea, well done!)
3. static factories -- "big" commands such as auto routines
4. subclassing command groups (with only a `super()` call; no overriding methods) -- "big" commands such as auto routines

Reply:
> Most of the articles (especially the command-based ones) are spreads of information, and the narrative/progression is only at the article level (ie articles about simpler topics are before articles about advanced topics).

Yes, but I'm explicitly not following that model here—I plan to write a more spec-sheet-y companion piece that's just "here's a good structure for intermediate robot code" (that might need some design discussion). Maybe I should put a bigger notice up at the top, but this is designed for programmers who want to do some legitimate thinking about what works best for them.

I'm not sure I agree with your categorization of which command group is best for which use case (which sort of indicates how this works better as a design _discussion_ rather than a prescriptive guide). 

For example, for single commands that have somewhat complicated logic, I personally would usually avoid subclassing `CommandBase` and instead use a `FunctionalCommand` in a static or non-static factory method. (I'd also try to just generally avoid complicated logic in commands, and try to achieve the functionality via command composition instead.) But again, this is a matter of personal taste—figuring out command composition is often more difficult for inexperienced teams then building a somewhat chunky command subclass with its own little state machine inside. 

I may take another crack at revising things, though, so watch this space.

Reply:
> I personally would usually avoid subclassing CommandBase and instead use a FunctionalCommand in a static or non-static factory method. 

Since you're implementing all four command lifecycle methods anyway, there's little difference in verbosity; writing a class allows you to more easily share state between the lifecycle methods. Having said all that, I do agree that it's a matter of personal preference.
As for the rest of that paragraph, I 100% agree.

---

> Yes, but I'm explicitly not following that model here—I plan to write a more spec-sheet-y companion piece that's just "here's a good structure for intermediate robot code" (that might need some design discussion). Maybe I should put a bigger notice up at the top, but this is designed for programmers who want to do some legitimate thinking about what works best for them.

I'm not sure what we're disagreeing here about. What I was thinking of for this doc is an explanation of each option with pros and cons, and leave it as an exercise to the reader to figure out what's appropriate and works best for their use case. I was not intending for this to be a "The *One* True Way".
However, concrete examples massively help to understand these abstract topics, and examples need to be based on experience--which is why it's open for discussion. I gave those examples because that's what matches my experience, I'm open to others' experience.